File: client/java-client/src/main/java/com/kotak/neo/client/rest/RestClient.java
package com.kotak.neo.client.rest;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.kotak.neo.client.Configuration;
import com.kotak.neo.client.exceptions.ApiException;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.stream.Collectors;

public class RestClient {
    private static final Gson GSON = new Gson();
    private final HttpClient http;
    public final Configuration config;

    public RestClient(Configuration config) {
        this.config = config;
        this.http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build();
    }

    public static class Response {
        public int status;
        public String text;
        public JsonElement data;
    }

    public Response request(String method, String url, Map<String, String> queryParams,
                            Map<String, String> headers, Object body) {
        method = method.toUpperCase();
        if (headers == null) headers = new java.util.HashMap<>();
        headers.putIfAbsent("Content-Type", "application/json");
        headers.putIfAbsent("User-Agent", "NeoTradeApi-java/1.0.0");

        if (queryParams != null && !queryParams.isEmpty()) {
            String qs = queryParams.entrySet().stream()
                    .map(e -> enc(e.getKey()) + "=" + enc(e.getValue()))
                    .collect(Collectors.joining("&"));
            url += (url.contains("?") ? "&" : "?") + qs;
        }

        HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(url));
        HttpRequest.BodyPublisher bodyPublisher = HttpRequest.BodyPublishers.noBody();

        String ct = headers.get("Content-Type");
        if (method.equals("POST") || method.equals("PUT") || method.equals("PATCH") || method.equals("DELETE")) {
            if (ct.toLowerCase().contains("json")) {
                String payload = body == null ? "" : GSON.toJson(body);
                bodyPublisher = HttpRequest.BodyPublishers.ofString(payload, StandardCharsets.UTF_8);
            } else if (ct.toLowerCase().contains("x-www-form-urlencoded")) {
                String payload = "";
                if (body != null) {
                    payload = "jData=" + enc(GSON.toJson(body));
                }
                bodyPublisher = HttpRequest.BodyPublishers.ofString(payload, StandardCharsets.UTF_8);
            } else {
                throw new ApiException(0, "Invalid Content-Type", ct);
            }
        }

        switch (method) {
            case "GET": b.GET(); break;
            case "DELETE": b.method("DELETE", bodyPublisher); break;
            default: b.method(method, bodyPublisher);
        }

        for (Map.Entry<String, String> h : headers.entrySet()) {
            // Restricted headers cannot be set directly; let the JDK add them.
            try { b.header(h.getKey(), h.getValue()); } catch (IllegalArgumentException ignored) { }
        }

        try {
            HttpResponse<String> resp = http.send(b.build(), HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            Response r = new Response();
            r.status = resp.statusCode();
            r.text = resp.body() == null ? "" : resp.body();
            try {
                r.data = r.text.isEmpty() ? new JsonObject() : JsonParser.parseString(r.text);
            } catch (Exception parseFail) {
                JsonObject obj = new JsonObject();
                obj.addProperty("raw", r.text);
                r.data = obj;
            }
            return r;
        } catch (IOException | InterruptedException e) {
            throw new ApiException(0, "http error", e.getMessage());
        }
    }

    private static String enc(String s) {
        return URLEncoder.encode(s == null ? "" : s, StandardCharsets.UTF_8);
    }
}

File: client/dotnet-client/NeoApi/Rest/RestClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Kotak.Neo.Exceptions;

namespace Kotak.Neo.Rest;

public class RestClient
{
    private readonly HttpClient _http;
    public Configuration Config { get; }

    public RestClient(Configuration config, HttpClient? http = null)
    {
        Config = config;
        _http = http ?? new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
        _http.DefaultRequestHeaders.UserAgent.ParseAdd("NeoTradeApi-dotnet/1.0.0");
    }

    public async Task<(int Status, JsonElement Data, string Text)> RequestAsync(
        string method,
        string url,
        IDictionary<string, string>? queryParams = null,
        IDictionary<string, string>? headers = null,
        object? body = null,
        CancellationToken ct = default)
    {
        headers ??= new Dictionary<string, string>();
        if (!headers.ContainsKey("Content-Type"))
            headers["Content-Type"] = "application/json";

        if (queryParams is { Count: > 0 })
        {
            var pairs = string.Join("&", queryParams.Select(kv =>
                $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));
            url += (url.Contains('?') ? "&" : "?") + pairs;
        }

        var request = new HttpRequestMessage(new HttpMethod(method.ToUpperInvariant()), url);
        var contentType = headers["Content-Type"];

        if (method is "POST" or "PUT" or "PATCH" or "DELETE")
        {
            if (contentType.Contains("json", StringComparison.OrdinalIgnoreCase))
            {
                var json = body is null ? "" : JsonSerializer.Serialize(body);
                request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            }
            else if (contentType.Contains("x-www-form-urlencoded", StringComparison.OrdinalIgnoreCase))
            {
                var pairs = new List<KeyValuePair<string, string>>();
                if (body is not null)
                {
                    var json = JsonSerializer.Serialize(body);
                    pairs.Add(new KeyValuePair<string, string>("jData", json));
                }
                request.Content = new FormUrlEncodedContent(pairs);
            }
            else
            {
                throw new ApiException(0, "Invalid Content-Type", contentType);
            }
        }

        foreach (var (k, v) in headers)
        {
            if (k.Equals("Content-Type", StringComparison.OrdinalIgnoreCase)) continue;
            request.Headers.TryAddWithoutValidation(k, v);
        }

        var resp = await _http.SendAsync(request, ct).ConfigureAwait(false);
        var text = await resp.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        JsonElement data;
        try
        {
            data = string.IsNullOrEmpty(text)
                ? JsonDocument.Parse("{}").RootElement.Clone()
                : JsonDocument.Parse(text).RootElement.Clone();
        }
        catch
        {
            using var doc = JsonDocument.Parse(JsonSerializer.Serialize(new { raw = text }));
            data = doc.RootElement.Clone();
        }
        return ((int)resp.StatusCode, data, text);
    }
}
